pydantic/main.py

```diff
@@ -592,6 +603,51 @@ def __repr_args__(self) -> _repr.ReprArgs:
             if not k.startswith('_') and (k not in self.model_fields or self.model_fields[k].repr)
         ]
 
+    def __class_getitem__(cls, item: typing.Union[type[Any], typing.Tuple[type[Any], ...]]) -> type[Any]:
+        def _cache_key(_params: Any) -> typing.Tuple[type[Any], Any, typing.Tuple[Any, ...]]:
+            return cls, _params, typing_extensions.get_args(_params)
+
+        cached = _generic_types_cache.get(_cache_key(item))
+        if cached is not None:
+            return cached
+
+        if not isinstance(item, tuple):
+            item = (item,)
+
+        if not is_generic_model(cls):
+            raise TypeError('Cannot parameterize a non-generic BaseModel subclass')
+
+        generic_alias = super().__class_getitem__(item)  # type: ignore[misc]
+
+        model_name = cls.__concrete_name__(item)
+        submodel = create_generic_submodel(model_name, cls)
+
+        typevars_map = dict(zip(submodel.__parameters__, item))
+        submodel.model_rebuild(force=True, typevars_map=typevars_map)
+        submodel.__parameters__ = generic_alias.__parameters__
+
+        _generic_types_cache[_cache_key(item)] = submodel
+        if len(item) == 1:
+            _generic_types_cache[_cache_key(item[0])] = submodel
+
+        return submodel
+
+    @classmethod
+    def __concrete_name__(cls: type[Any], params: typing.Tuple[type[Any], ...]) -> str:
+        """Compute class name for child classes.
+
+        :param params: Tuple of types the class . Given a generic class
+            `Model` with 2 type variables and a concrete model `Model[str, int]`,
+            the value `(str, int)` would be passed to `params`.
+        :return: String representing a the new class where `params` are
+            passed to `cls` as type variables.
+
+        This method can be overridden to achieve a custom naming scheme for GenericModels.
```

Comment:
Obviously need to update these docs, but I wasn't sure how you were going to feel about keeping `__concrete_name__` as a method on `BaseModel`. Having the ability to modify the name-generation logic feels like a significant benefit, I just feel a bit averse to polluting the `BaseModel` namespace.

Reply:
new rule for v2 to try and be better python citizens_ we want all dunder names to start with `__pydantic` (or at least include pydanitc, e.g. `__get_pydantic...`).

Should we use `__pydantic_concrete_name__` here?

(I assume `__concrete_name__` is not defined in the stdlib, or a pepe somewhere?)

Or, since this is not really private or magic, maybe we should just use `model_concrete_name`???

Reply:
I I think I prefer a public name, e.g. `model_concrete_name`.

Reply:
@samuelcolvin I'm changing it to `model_concrete_name` for now, but one other alternative I'd like to propose: for public methods, using `model__` as a prefix, rather than `model_` (so `model__concrete_name` instead of `model_concrete_name`; similarly `model__json`, etc.).

This feels significantly less likely to run into collisions with user-desired fields (or methods), and to my eye almost *more* readable (as I can immediately tell it is a pydantic model method, since nothing else uses a format like that), with only one extra character. And it retains the various benefits of the method name being pseudo-public (e.g., IDEs including it at the top of autocomplete, etc.).

If you consider this an already-finalized decision, or just hate this proposal, let's just pretend this comment never happened :).

Reply:
let's still with `model_concrete_name` here for consistency.

Happy to consider `model__` prefix, probably worth discussing on #4915 or in a new issue.

I dislike all proposals in this area, including my own - there's no elegant way around this problem.